lighthouse-core/gather/gatherers/seo/font-size.js

```diff
@@ -44,58 +48,130 @@ function nodeInBody(node) {
  * Get list of all nodes from the document body.
  *
  * @param {Driver} driver
- * @returns {Array<LH.Artifacts.FontSize.DomNodeWithParent>}
+ * @returns {Promise<Array<LH.Artifacts.FontSize.DomNodeWithParent>>}
  */
-function getAllNodesFromBody(driver) {
-  return driver.getNodesInDocument()
-    .then(nodes => {
-      const nodeMap = new Map();
-      nodes.forEach(node => nodeMap.set(node.nodeId, node));
-      nodes.forEach(node => node.parentNode = nodeMap.get(node.parentId));
-      return nodes.filter(nodeInBody);
-    });
+async function getAllNodesFromBody(driver) {
+  const nodes = /** @type {DomNodeMaybeWithParent[]} */ (await driver.getNodesInDocument());
+  /** @type {Map<number|undefined, LH.Artifacts.FontSize.DomNodeMaybeWithParent>} */
+  const nodeMap = new Map();
+  nodes.forEach(node => nodeMap.set(node.nodeId, node));
+  nodes.forEach(node => (node.parentNode = nodeMap.get(node.parentId)));
+  return nodes.filter(nodeInBody);
+}
+
+/**
+ * @param {LH.Crdp.CSS.CSSStyle} [style]
+ * @return {boolean}
+ */
+function hasFontSizeDeclaration(style) {
+  return !!style && !!style.cssProperties.find(({name}) => name === FONT_SIZE_PROPERTY_NAME);
 }
 
 /**
- * Returns effective CSS rule for given CSS property
+ * Computes the CSS specificity of a given selector, i.e. #id > .class > div
  *
- * @param {string} property CSS property name
- * @param {LH.Crdp.DOM.Node} node
+ * @see https://developer.mozilla.org/en-US/docs/Web/CSS/Specificity
+ * @see https://www.smashingmagazine.com/2010/04/css-specificity-and-inheritance/
+ *
+ * @param {string} selector
+ * @return {number}
+ */
+function computeSelectorSpecificity(selector) {
+  const tokens = selector.split(' ');
+
+  let numIDs = 0;
+  let numClasses = 0;
+  let numTypes = 0;
+
+  for (const token of tokens) {
+    const ids = token.match(/\b#[a-z0-9]+/g) || [];
+    const classes = token.match(/\b\.[a-z0-9]+/g) || [];
+    const types = token.match(/^[a-z]+/) ? [1] : [];
+    numIDs += ids.length;
+    numClasses += classes.length;
+    numTypes += types.length;
+  }
+
+  return Math.min(9, numIDs) * 100 + Math.min(9, numClasses) * 10 + Math.min(9, numTypes);
```

Comment:
I've ignored `!important` so far, think it's worth handling? I'll at least comment it's absence

Reply:
> think it's worth handling?

no idea :) I have used it in anger more than once, so if you really really mean to override a font-size with it, it would be annoying if lighthouse didn't recognize it.

Is it non-trivial to support?

Reply:
> Is it non-trivial to support?

Kinda, we basically have to do this entire thing twice, once looking for !important and once not. The good news is that LH will still recognize the percentage of text going down. If you `!important`'d a valid font size, the risk is that you `!important` an also invalid font-size and we point to a rule that isn't really winning.

Reply:
ha, I guess I hadn't completely thought through how this gatherer is working.

Of course we already have the computed style, so the css parsing is just to get the rule that caused that style. It's complicated enough that I had forgotten that by this point 😳 I'll add a comment on `analyzeFailingNodes`